netbox/custom_fields.go

```diff
@@ -16,10 +25,140 @@ var customFieldsSchema = &schema.Schema{
+				Required:    true,
+				Description: "Name of the existing custom field.",
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{"text", "integer", "boolean",
+					"date", "url", "selection", "multiple"}, false),
+				Description: "Type of the existing custom field (text, integer, boolean, url, selection, multiple).",
+			},
+			"value": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Value of the existing custom field.",
+			},
+		},
+	},
+	Description: "Existing custom fields to associate to this prefix (ipam module).",
+}
+
 func getCustomFields(cf interface{}) map[string]interface{} {
 	cfm, ok := cf.(map[string]interface{})
 	if !ok || len(cfm) == 0 {
 		return nil
 	}
 	return cfm
 }
+
+func convertArrayInterfaceString(arrayInterface []interface{}) string {
+	var arrayString []string
+
+	for _, item := range arrayInterface {
+		switch v := item.(type) {
+		case string:
+			arrayString = append(arrayString, v)
+		case int:
+			strV := strconv.FormatInt(int64(v), 10)
+			arrayString = append(arrayString, strV)
+		}
+
+	}
+
+	sort.Strings(arrayString)
+	result := strings.Join(arrayString, ",")
+
+	return result
+}
+
+// Pick the custom fields in the state file and update values with data from API
+func updateCustomFieldsFromAPI(stateCustomFields, customFields interface{}) []map[string]string {
+	var tfCms []map[string]string
+
+	switch t := customFields.(type) {
+	case map[string]interface{}:
+		for _, stateCustomField := range stateCustomFields.([]interface{}) {
+			for key, value := range t {
+				if stateCustomField.(map[string]interface{})["name"].(string) == key {
+					var strValue string
+					cm := map[string]string{}
+					cm["name"] = key
+					cm["type"] = stateCustomField.(map[string]interface{})["type"].(string)
+
+					if value != nil {
+						switch v := value.(type) {
+						case []interface{}:
+							strValue = convertArrayInterfaceString(v)
+						default:
+							strValue = fmt.Sprintf("%v", v)
+						}
+
+						if strValue == "1" && cm["type"] == CustomFieldBoolean {
+							strValue = "true"
+						} else if strValue == "0" && cm["type"] == CustomFieldBoolean {
+							strValue = "false"
+						}
+
+						cm["value"] = strValue
+					} else {
+						cm["value"] = ""
+					}
+
+					tfCms = append(tfCms, cm)
+				}
+			}
+		}
+	}
+
+	return tfCms
+}
+
+// Convert custom field regarding his type
+func convertCustomFieldsFromTerraformToAPI(stateCustomFields []interface{}, customFields []interface{}) map[string]interface{} {
+	toReturn := make(map[string]interface{})
+
+	for _, stateCf := range stateCustomFields {
+		stateCustomField := stateCf.(map[string]interface{})
+		toReturn[stateCustomField["name"].(string)] = nil
+	}
+
+	for _, cf := range customFields {
+		customField := cf.(map[string]interface{})
+
+		cfName := customField["name"].(string)
+		cfType := customField["type"].(string)
```

Comment:
Regarding my previous comment, this datapoint should exists in the state to compare a diff, but should be discovered by the netbox API instead of a declaration in HCL
By the way, the provider should keep in memory cache the custom_fields objects, because the can be associated to multiple objects types, instead of asking them multiple times to netbox.

Reply:
Thanks for suggestion, sounds reasonable. Could you please show me an example? I am very bad in golang :( Otherwise I can try, but will take me some time.

Reply:
Sorry, I'm worst than you in Golang from scratch! ^^
That's more an architectural advice to avoid performance issues and unexpected side effects over the time in case of manual changes in netbox.
@fbreckle or someone else, have you some advices ?